modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/SqlPlanHistoryIntegrationTest.java

```diff
@@ -0,0 +1,718 @@
+        assumeFalse("There is no 'local query' parameter for JDBC queries", loc);
+
+        try (
+            Connection conn = GridTestUtils.connect(queryNode(), null);
+            Statement stmt = conn.createStatement()
+        ) {
+            if (!isFullyFetched)
+                stmt.setFetchSize(1);
+
+            ResultSet rs = stmt.executeQuery(qry);
+
+            assertTrue(rs.next());
+        }
+    }
+
+    /**
+     * @param qry Query.
+     * @param cacheName Cache name.
+     */
+    public void cacheQuery(Query qry, String cacheName) {
+        IgniteCache<Integer, String> cache = queryNode().getOrCreateCache(cacheName);
+
+        if (isFullyFetched)
+            assertFalse(cache.query(qry).getAll().isEmpty());
+        else {
+            qry.setPageSize(1);
+
+            assertNotNull(cache.query(qry).iterator().next());
+        }
+    }
+
+    /**
+     * @param qrysInfo DML commands info (queries, simple query flag).
+     */
+    public void runJdbcDml(IgniteBiTuple<List<String>, Boolean> qrysInfo) {
+        assumeFalse("There is no 'local query' parameter for JDBC queries", loc);
+
+        executeDml(qrysInfo, (cmds) -> {
+            try (
+                Connection conn = GridTestUtils.connect(queryNode(), null);
+                Statement stmt = conn.createStatement()
+            ) {
+                for (String cmd : cmds)
+                    stmt.execute(cmd);
+            }
+            catch (SQLException e) {
+                new RuntimeException(e);
+            }
+        });
+    }
+
+    /**
+     * @param qrysInfo DML commands info (queries, simple query flag).
+     */
+    public void runSqlFieldsDml(IgniteBiTuple<List<String>, Boolean> qrysInfo) {
+        executeDml(qrysInfo, (cmds) -> {
+            IgniteCache<Integer, String> cache = queryNode().getOrCreateCache("A");
+
+            cmds.forEach(cmd -> cache.query(new SqlFieldsQuery(cmd).setLocal(loc)));
+        });
+    }
+
+    /**
+     * @param qrysInfo DML commands info (queries, simple query flag).
+     * @param task Task to be executed.
+     */
+    public void executeDml(IgniteBiTuple<List<String>, Boolean> qrysInfo, Consumer<List<String>> task) {
+        assumeFalse("There is no lazy mode for DML operations", !isFullyFetched);
+
+        List<String> cmds = qrysInfo.get1();
+
+        task.accept(cmds);
+
+        checkSqlPlanHistoryDml(3, qrysInfo.get2());
+    }
+
+    /** Returns current SQL plan history on the query node. */
+    public Map<SqlPlan, Long> getSqlPlanHistory() {
+        return queryNode().context().query().runningQueryManager().planHistoryTracker().sqlPlanHistory();
+    }
+
+    /**
+     * Checks SQL plan history for select queries.
+     *
+     * @param size Number of SQL plan entries expected to be in the history.
+     */
+    public void checkSqlPlanHistory(int size) {
+        if (!isReducePhase && isClient && sqlEngine == IndexingQueryEngineConfiguration.ENGINE_NAME)
```

Comment:
It's strange that we don't have plans in history for regular queries for client node and H2 engine. Do we need to check H2 & client at all? Map/reduce queries are only applicable to H2 engine and can be reproduced with two server nodes (even without distributed joins, for example with aggregation)

Reply:
The `isClient=TRUE` parameter has been deleted from the H2 engine tests and the map/reduce query test has been simplified.